### PR TITLE
Add pipeline configuration for cleaning up upstream branches

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Add Cleanup pipeline.
- The intention is to run this on a schedule. The first step here is for cleaning up `eng sommon sync` branches.

### CODE REVIEW

File: eng/common/scripts/Delete-RemoteBranches.ps1 [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+pushd $WorkingDirectory
+git clone https://github.com/$RepoOwner/$RepoName.git
+pushd $RepoName
+$syncBranches = git branch -r --list origin/$BranchPrefix* | % { $_ -replace "origin/", "" }
+
+LogDebug "Operating on Repo [ $RepoName ]"
+foreach ($branch in $syncBranches)
+{
+  try {
+    $branchName = $branch.Trim()
```
Comment (weshaggard) [inline eng/common/scripts/Delete-RemoteBranches.ps1:20]: If it is interesting to trim then I would do it above when producing the list of branches.

File: eng/pipelines/templates/jobs/tools-cleanup.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      - ${{ each repo in parameters.Repos }}:
+        - task: PowerShell@2
+          displayName: Clean Up Sync Common Branches
+          condition: succeeded()
+          inputs:
+            pwsh: true
+            workingDirectory: $(System.DefaultWorkingDirectory)
+            filePath: $(System.DefaultWorkingDirectory)/eng/common/scripts/Delete-RemoteBranches.ps1
+            arguments: >
+              -RepoName ${{ repo }}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/tools-cleanup.yml:29]: Looks like you are not passing RepoOwner. I assume you plan to pass azure-sdk as these are cleaning up branches in the forks. 

We should also consider doing a similar branch clean-up for the increment version branches.

**Review by weshaggard — APPROVED** [review]
Add a couple comments but looks good overally.

File: eng/common/scripts/Delete-RemoteBranches.ps1 [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  $WorkingDirectory,
+  $AuthToken
+)
+
+. "${PSScriptRoot}\common.ps1"
+
+pushd $WorkingDirectory
+git clone https://github.com/$RepoOwner/$RepoName.git
+pushd $RepoName
+$syncBranches = git branch -r --list origin/$BranchPrefix* | % { $_ -replace "origin/", "" }
```
Comment (mitchdenny) [inline eng/common/scripts/Delete-RemoteBranches.ps1:14]: Can you just use this API to query for branches and delete them without having to clone the repo?

https://docs.github.com/en/free-pro-team@latest/rest/reference/git#delete-a-reference

Should be much faster.

Comment (chidozieononiwu) [reply eng/common/scripts/Delete-RemoteBranches.ps1]: Thanks, I looked everywhere for this. Weird I could not find it.

File: eng/common/scripts/Delete-RemoteBranches.ps1 [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  "bvranch $branch"
+  "PR COunt $($pullRequests.Count)"
+
+
+  if ($pullRequests.Count -eq 0)
+  {
+    LogDebug "Branch [ $branchName ] in repo [ $RepoName ] has no associated Pull Request. Deleting Branch"
+    try{
+      Delete-References -RepoOwner $RepoOwner -RepoName $RepoName -Ref ($branch.Remove(0,5)) -AuthToken $AuthToken
```
Comment (mitchdenny) [inline eng/common/scripts/Delete-RemoteBranches.ps1:41]: In addition to deleting the reference, should we also delete the pull request itself or will that happen automatically?

**Review by mitchdenny — APPROVED** [review]
Just an open question about whether we need to delete the pull request as well, but other than that looks good.

File: eng/common/scripts/Delete-RemoteBranches.ps1 (hunk 4 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Delete-RemoteBranches.ps1:41]: Am not sure what you mean? Do we want to delete the pull request. Isn't it better to keep that after merging it.

Comment (mitchdenny) [reply eng/common/scripts/Delete-RemoteBranches.ps1:37]: Sorry, I was thinking that this was a a clean-up just incase we wanted to abort. I take back my feedback ;)

**Comment by weshaggard** [comment]
@chidozieononiwu one side-effect of using the auto-merge tag we end up loosing some of the commit history when we it squashes and merges. See https://github.com/Azure/azure-sdk-for-c/commit/639007a840deda72c1492df843c6e482fe85b6e0 which is the auto-merged commit for https://github.com/Azure/azure-sdk-for-c/pull/1426 and compare that to the manually merged commit https://github.com/Azure/azure-sdk-for-python/pull/14482 for https://github.com/Azure/azure-sdk-for-python/pull/14482. 

The auto-merge will use the issue description as the commit message which is by design. So I guess we should decided if that is OK for our workflow or not. On one hand I kind of like it because it has a direct link to the tools PR but on the other we do lose the commit messages from the tools changes which does force people to always click through to the tools PR. 

@mitchdenny what is your thoughts on what should happen here?

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 5]
```diff
@@ -105,6 +123,27 @@ function List-PullRequests {
   return Invoke-GitHubAPIGet -apiURI $uri
 }
 
+# 
+<#
+.PARAMETER Ref
+Ref to search for
+Pass 'heads/<branchame> ,tags/<tag name>, or nothing
+#>
+function List-References {
```
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:132]: I think we should have a better name for this function as List-References without context is hard to understand as it is very ambiguous with other things. For example reference means something in PS as well. Perhaps `List-GithubSourceReferences` or something like that might be a little clearer.

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 6]
```diff
@@ -105,6 +123,27 @@ function List-PullRequests {
...
+  param (
+    [Parameter(Mandatory = $true)]
+    $RepoOwner,
+    [Parameter(Mandatory = $true)]
+    $RepoName,
+    $Ref
```
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:138]: At some point we might also need to pass a token here if we ever call this on a private repo for example.

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 7]
```diff
@@ -229,4 +268,27 @@ function Update-Issue {
   }
 
   return Invoke-GitHubAPIPatch -apiURI $uri -body $parameters -token $AuthToken
+}
+
+function Delete-References {
```
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:273]: Perhaps a better name would be `Delete-GithubSourceReference`.

File: eng/common/scripts/Delete-RemoteBranches.ps1 [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    LogDebug "Operating on branch [ $branchName ]"
+    $pullRequests = List-PullRequests -RepoOwner $RepoOwner -RepoName $RepoName -head $head
+  }
+  catch
+  {
+    LogError "List-PullRequests failed with exception:`n$_"
+    exit 1
+  }
+
+  if ($pullRequests.Count -eq 0)
```
Comment (weshaggard) [inline eng/common/scripts/Delete-RemoteBranches.ps1:33]: Should we log cases where we find branches that have a pull request and list the branch and what open pull requests there are still for it?

**Pull request merged**